Tests/TestLinkDecorator.swift

```diff
@@ -0,0 +1,180 @@
+    let epoch = "\\d{13}"
+    
+    let replacements = [".", "/", "?"]
+    func matches(for regex: String, in text: String) {
+        var regex = "^\(regex)$"
+        
+        do {
+            for replacement in replacements {
+                regex = regex.replacingOccurrences(of: replacement, with: "\\" + replacement)
+            }
+            let pattern = try NSRegularExpression(pattern: regex)
+            let nsString = text as NSString
+            let results = pattern.matches(in: text, range: NSRange(location: 0, length: nsString.length))
+            let fullMatch = results.map { nsString.substring(with: $0.range)}
+            if (fullMatch.count == 0) {
+                XCTFail("URL does not match pattern:\n\(text)\n\(regex)")
+            }
+            XCTAssertEqual(fullMatch.count, 1)
+        } catch let error {
+            print("invalid regex: \(error.localizedDescription)")
+        }
+    }
+    
+    func testParameterConfiguration() {
+        let tracker = getTracker()
+        let _ = tracker.track(ScreenView(name: "test"))
+        
+        let link = URL(string: "https://example.com")!
+        let userId = tracker.session!.userId!
+        let sessionId = tracker.session!.sessionId!
+        let subjectUserId = tracker.subject!.userId!.toBase64()
+        let appId = tracker.appId.toBase64()
+        let platform = devicePlatformToString(tracker.devicePlatform)!
+        let reason = "reason"
+        let reasonb64 = reason.toBase64()
+        
+        let testCases = [
+            // All false
+            (
+                config: CrossDeviceParameterConfiguration(sessionId: false, sourceId: false),
+                expected: "https://example.com?_sp=\(userId).\(epoch)"
+            )
+            ,
+            // Default
+            (
+                config: CrossDeviceParameterConfiguration(),
+                expected: "https://example.com?_sp=\(userId).\(epoch).\(sessionId)..\(appId)"
+            ),
+            (
+                config: CrossDeviceParameterConfiguration(subjectUserId: true),
+                expected: "https://example.com?_sp=\(userId).\(epoch).\(sessionId).\(subjectUserId).\(appId)"
+            ),
+            (
+                config: CrossDeviceParameterConfiguration(subjectUserId: true, sourcePlatform: true),
+                expected: "https://example.com?_sp=\(userId).\(epoch).\(sessionId).\(subjectUserId).\(appId).\(platform)"
+            ),
+            (
+                config: CrossDeviceParameterConfiguration(subjectUserId: true, sourcePlatform: true, reason: reason),
+                expected: "https://example.com?_sp=\(userId).\(epoch).\(sessionId).\(subjectUserId).\(appId).\(platform).\(reasonb64)"
+            ),
+            (
+                config: CrossDeviceParameterConfiguration(sessionId: false, sourceId: false, reason: reason),
+                expected: "https://example.com?_sp=\(userId).\(epoch).....\(reasonb64)"
```

Comment:
If this ends up in a regular expression, you might need to escape the dots and perhaps also the ?

Reply:
They are escaped [within](https://github.com/snowplow/snowplow-ios-tracker/pull/829/files#diff-3f8446682cdd6702304d274c3f20f4ae0af41b4d41322380183e4ced68a4ebacR26) the `matches` function to make the expected strings less busy, but maybe that's a bit confusing...